File: lab11/iterator.kt
import java.util.*

// Iterator
interface FibonacciIterator {
  fun hasNext(): Boolean
  fun next(): ULong
}

// Aggregate
interface FibonacciCollection {
  fun createIterator(): FibonacciIterator
}

// Concrete Aggregate
class FibonacciSequence(count: Int) : FibonacciCollection {

  val sequence: ULongArray

  init {
    sequence = ULongArray(count)

    sequence[0] = 0UL
    if(count > 1)
      sequence[1] = 1UL

    for(i in 2 until count)
      sequence[i] = sequence[i - 1] + sequence[i - 2]
  }

  fun getNumber(index: Int): ULong {
    return sequence[index]
  }

  fun size(): Int {
    return sequence.size
  }

  override fun createIterator(): FibonacciIterator {
    return SequenceIterator(this)
  }
}

// Concrete Iterator
class SequenceIterator(seq: FibonacciSequence) : FibonacciIterator {

  val sequence: FibonacciSequence
  var index = 0

  init {
    sequence = seq
  }

  override fun hasNext(): Boolean {
    return index < sequence.size()
  }

  override fun next(): ULong {
    return sequence.getNumber(index++)
  }
}

fun main() {
  val fib: FibonacciCollection = FibonacciSequence(20)
  val it: FibonacciIterator = fib.createIterator()

  println("Iterating Fibonacci numbers...\n")

  while(it.hasNext()) {
    print("${it.next()} ")
  }
  println()

  return
}

File: lab11/iterator.cpp
#include <iostream>
#include <vector>

using namespace std;

// Iterator
class FibonacciIterator {
public:
  virtual bool hasNext() = 0;
  virtual unsigned long long next() = 0;
  virtual ~FibonacciIterator() = default;
};

// Aggregate
class FibonacciCollection {
public:
  virtual FibonacciIterator* createIterator() = 0;
  virtual ~FibonacciCollection() = default;
};

// Concrete Aggregate
class FibonacciSequence : public FibonacciCollection {
private:
  vector<unsigned long long> sequence;

public:
  FibonacciSequence(int count) {
    sequence.resize(count);

    sequence[0] = 0;
    if(count > 1)
      sequence[1] = 1;

    for(int i = 2; i < count; i++)
      sequence[i] = sequence[i - 1] + sequence[i - 2];
  }

  unsigned long long getNumber(int index) const {
    return sequence[index];
  }

  int size() const {
    return sequence.size();
  }

  FibonacciIterator* createIterator() override;
};

// Concrete Iterator
class SequenceIterator : public FibonacciIterator {
private:
  FibonacciSequence* sequence;
  int index = 0;

public:
  SequenceIterator(FibonacciSequence* seq)
    : sequence(seq) {}

  bool hasNext() override {
    return index < sequence->size();
  }

  unsigned long long next() override {
    return sequence->getNumber(index++);
  }
};

FibonacciIterator* FibonacciSequence::createIterator() {
  return new SequenceIterator(this);
}

int main() {
  FibonacciCollection* fib = new FibonacciSequence(20);
  FibonacciIterator* it = fib->createIterator();

  cout << "Iterating Fibonacci numbers...\n";

  while(it->hasNext()) {
    cout << it->next() << " ";
  }
  cout << '\n';

  delete it;
  delete fib;
  return 0;
}
